python/ray/experimental/multiprocessing/pool.py

```diff
@@ -0,0 +1,488 @@
+        return AsyncResult(
+            [object_id], callback, error_callback, single_result=True)
+
+    def map(self, func, iterable, chunksize=None):
+        self._check_running()
+        return self._map_async(
+            func, iterable, chunksize=chunksize, unpack_args=False).get()
+
+    def map_async(self,
+                  func,
+                  iterable,
+                  chunksize=None,
+                  callback=None,
+                  error_callback=None):
+        self._check_running()
+        return self._map_async(
+            func,
+            iterable,
+            chunksize=chunksize,
+            unpack_args=False,
+            callback=callback,
+            error_callback=error_callback)
+
+    def starmap(self, func, iterable, chunksize=None):
+        self._check_running()
+        return self._map_async(
+            func, iterable, chunksize=chunksize, unpack_args=True).get()
+
+    def starmap_async(self, func, iterable, callback=None,
+                      error_callback=None):
+        self._check_running()
+        return self._map_async(
+            func,
+            iterable,
+            unpack_args=True,
+            callback=callback,
+            error_callback=error_callback)
+
+    def _map_async(self,
+                   func,
+                   iterable,
+                   chunksize=None,
+                   unpack_args=False,
+                   callback=None,
+                   error_callback=None):
+        object_ids = self._chunk_and_run(
+            func, iterable, chunksize=chunksize, unpack_args=unpack_args)
+        return AsyncResult(object_ids, callback, error_callback)
+
+    def _calculate_chunksize(self, iterable):
+        chunksize, extra = divmod(len(iterable), len(self._actor_pool) * 4)
+        if extra:
+            chunksize += 1
+        return chunksize
+
+    def _submit_chunk(self,
+                      func,
+                      iterator,
+                      chunksize,
+                      actor_index,
+                      unpack_args=False):
+        chunk = []
+        while len(chunk) < chunksize:
+            try:
+                args = iterator.__next__()
+                if not unpack_args:
+                    args = (args, )
+                chunk.append((args, {}))
+            except StopIteration:
+                break
+        if len(chunk) == 0:
+            return None
+        return self._run_batch(actor_index, func, chunk)
+
+    def _chunk_and_run(self, func, iterable, chunksize=None,
+                       unpack_args=False):
+        if not hasattr(iterable, "__len__"):
+            iterable = [iterable]
+
+        if chunksize is None:
+            chunksize = self._calculate_chunksize(iterable)
+
+        iterator = iter(iterable)
+        chunk_object_ids = []
+        while len(chunk_object_ids) * chunksize < len(iterable):
+            actor_index = len(chunk_object_ids) % len(self._actor_pool)
```

Comment:
Is `actor_index` needed as an extra variable? If not, just put the value in the call below. If it's unclear from context (I don't know), you can use a keyword argument.